source-mysql/replication.go

```diff
@@ -0,0 +1,214 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/estuary/connectors/sqlcapture"
+	"github.com/go-mysql-org/go-mysql/mysql"
+	"github.com/go-mysql-org/go-mysql/replication"
+	"github.com/sirupsen/logrus"
+)
+
+// replicationBufferSize controls how many change events can be buffered in the
+// replicationStream before it stops receiving further events from MySQL.
+// In normal use it's a constant, it's just a variable so that tests are more
+// likely to exercise blocking sends and backpressure.
+var replicationBufferSize = 1024
+
+func (db *mysqlDatabase) StartReplication(ctx context.Context, startCursor string) (sqlcapture.ReplicationStream, error) {
+	var host, port, err = splitHostPort(db.config.Address)
+	if err != nil {
+		return nil, fmt.Errorf("invalid mysql address: %w", err)
+	}
+	var syncer = replication.NewBinlogSyncer(replication.BinlogSyncerConfig{
+		ServerID: uint32(db.config.ServerID),
+		Flavor:   "mysql", // TODO(wgd): See what happens if we change this and run against MariaDB?
+		Host:     host,
+		Port:     uint16(port),
+		User:     db.config.User,
+		Password: db.config.Pass,
+	})
+
+	var pos mysql.Position
+	if startCursor != "" {
+		var binlogName, binlogPos, err = splitHostPort(startCursor)
```

Comment:
Yeah, the backstory is pretty much that I already had the `splitHostPort` function written to handle the address-splitting a few lines above here, and then when I added the resume cursor logic it needed the exact same behavior, but I also couldn't think of a better name for the function.

On reflection I think the right answer is actually to duplicate a few lines of code and define a second `splitCursor` function, because that lets me reword the error messages for each function to be more contextually useful.